units/en/unit2/hands-on.mdx

```diff
@@ -0,0 +1,1074 @@
+# Each seed has a specific starting state
+
+# Environment parameters
+env_id = "Taxi-v3"  # Name of the environment
+max_steps = 99  # Max steps per episode
+gamma = 0.95  # Discounting rate
+
+# Exploration parameters
+max_epsilon = 1.0  # Exploration probability at start
+min_epsilon = 0.05  # Minimum exploration probability
+decay_rate = 0.005  # Exponential decay rate for exploration prob
+```
+
+## Train our Q-Learning agent 🏃
+
+```python
+Qtable_taxi = train(n_training_episodes, min_epsilon, max_epsilon, decay_rate, env, max_steps, Qtable_taxi)
+```
+
+```python
+Qtable_taxi
+```
+
+## Create a model dictionary 💾 and publish our trained model on the Hub 🔥
+- We create a model dictionary that will contain all the training hyperparameters for reproducibility and the Q-Table.
+
+
+```python
+model = {
+    "env_id": env_id,
+    "max_steps": max_steps,
+    "n_training_episodes": n_training_episodes,
+    "n_eval_episodes": n_eval_episodes,
+    "eval_seed": eval_seed,
+    "learning_rate": learning_rate,
+    "gamma": gamma,
+    "max_epsilon": max_epsilon,
+    "min_epsilon": min_epsilon,
+    "decay_rate": decay_rate,
+    "qtable": Qtable_taxi,
+}
+```
+
+```python
+username = ""  # FILL THIS
+repo_name = "q-Taxi-v3"
+push_to_hub(repo_id=f"{username}/{repo_name}", model=model, env=env)
+```
+
+Now that's on the Hub, you can compare the results of your Taxi-v3 with your classmates using the leaderboard 🏆 👉 https://huggingface.co/spaces/huggingface-projects/Deep-Reinforcement-Learning-Leaderboard
+
+<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/notebooks/unit2/taxi-leaderboard.png" alt="Taxi Leaderboard">
+
+# Part 3: Load from Hub 🔽
+
+What's amazing with Hugging Face Hub 🤗 is that you can easily load powerful models from the community.
+
+Loading a saved model from the Hub is really easy:
+
+1. You go https://huggingface.co/models?other=q-learning to see the list of all the q-learning saved models.
+2. You select one and copy its repo_id
+
+<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/notebooks/unit2/copy-id.png" alt="Copy id">
+
+3. Then we just need to use `load_from_hub` with:
+- The repo_id
+- The filename: the saved model inside the repo.
+
+#### Do not modify this code
+
+```python
+from urllib.error import HTTPError
+
+from huggingface_hub import hf_hub_download
+
+
+def load_from_hub(repo_id: str, filename: str) -> str:
+    """
+    Download a model from Hugging Face Hub.
+    :param repo_id: id of the model repository from the Hugging Face Hub
+    :param filename: name of the model zip file from the repository
+    """
+    try:
+        from huggingface_hub import cached_download, hf_hub_url
+    except ImportError:
+        raise ImportError(
+            "You need to install huggingface_hub to use `load_from_hub`. "
+            "See https://pypi.org/project/huggingface-hub/ for installation."
+        )
```

Comment:
We installed it at the top, so this is not needed. These imports are also not used
```suggestion
```

Reply:
:heavy_check_mark: